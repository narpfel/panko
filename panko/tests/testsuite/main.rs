#![feature(exit_status_error)]

use std::borrow::Borrow;
use std::io;
use std::path::Path;
use std::path::PathBuf;
use std::process::Command;
use std::process::ExitStatus;
use std::process::Output;

use insta_cmd::assert_cmd_snapshot;
use insta_cmd::get_cargo_bin;
use itertools::Itertools as _;
use regex::Captures;
use regex::Regex;
use rstest::rstest;

trait CaptureOutputForLibtest {
    fn status_with_captured_output(&mut self) -> io::Result<ExitStatus>;
}

impl CaptureOutputForLibtest for Command {
    fn status_with_captured_output(&mut self) -> io::Result<ExitStatus> {
        let output = self.output()?;
        if !output.stdout.is_empty() {
            println!("{}", String::from_utf8_lossy(&output.stdout));
        }
        if !output.stderr.is_empty() {
            eprintln!("{}", String::from_utf8_lossy(&output.stderr));
        }
        Ok(output.status)
    }
}

fn relative_to(path: &Path, target: impl AsRef<Path>) -> &Path {
    path.strip_prefix(target.as_ref()).unwrap()
}

#[rstest]
#[case::scopes("scope", "scopes")]
#[case::typeck("typeck", "typeck")]
#[case::layout("layout", "layout")]
fn test(
    #[case] snapshot_name_prefix: &str,
    #[case] step: &str,
    #[files("tests/cases/**/test_*.c")] filename: PathBuf,
) {
    let filename = relative_to(
        &filename,
        Path::new(env!("CARGO_MANIFEST_DIR")).parent().unwrap(),
    );
    assert_cmd_snapshot!(
        format!("{snapshot_name_prefix}-{}", filename.display()),
        Command::new(get_cargo_bin("panko"))
            .current_dir("..")
            .arg(format!("--print={step}"))
            .arg(format!("--stop-after={step}"))
            .arg(filename),
    );
}

#[rstest]
fn execute_test(#[files("tests/cases/execute/**/test_*.c")] filename: PathBuf) {
    let source = std::fs::read_to_string(&filename).unwrap();

    let expected_return_code_re =
        Regex::new(r"(?m)^// \[\[return: (?P<return_code>.*?)\]\]$").unwrap();

    let expected_return_codes = expected_return_code_re.captures_iter(&source).map(|captures| {
        let return_code = captures.name("return_code").unwrap().as_str();
        return_code.parse::<i32>().unwrap_or_else(|err| {
            panic!("while parsing `return_code` in {expected_return_code_re:?}:\n{err:?} in capture {return_code:?}")
        })
    }).collect_vec();
    assert!(
        expected_return_codes.len() <= 1,
        "test source invalid: more than one return code set: {expected_return_codes:?}",
    );
    let expected_return_code = expected_return_codes.first().copied().unwrap_or(0);

    let expected_print_re = Regex::new(r"(?m)^// \[\[print: (?P<output>.*?)\]\]$").unwrap();
    let expected_output: String = expected_print_re
        .captures_iter(&source)
        .map(|captures| captures.name("output").unwrap().as_str().to_string() + "\n")
        .collect();

    let cmdline_arguments_re = Regex::new(r"(?m)^// \[\[arg: (?P<arg>.*?)\]\]$").unwrap();
    let expand_escape_sequences_re = Regex::new(r"\\[\\n]").unwrap();

    let cmdline_arguments = cmdline_arguments_re
        .captures_iter(&source)
        .map(|captures| {
            expand_escape_sequences_re.replace_all(
                captures.name("arg").unwrap().as_str(),
                |captures: &Captures| match captures.get(0).unwrap().as_str() {
                    r"\\" => "\\",
                    r"\n" => "\n",
                    _ => unreachable!(),
                },
            )
        })
        .collect_vec();

    let filename = relative_to(
        &filename,
        Path::new(env!("CARGO_MANIFEST_DIR")).parent().unwrap(),
    );

    let output_dir = tempfile::tempdir().unwrap();
    let executable_filename = output_dir
        .path()
        .join(filename.file_name().unwrap())
        .with_extension("");

    Command::new(get_cargo_bin("panko"))
        .current_dir("..")
        .arg(filename)
        .arg("-o")
        .arg(&executable_filename)
        .status_with_captured_output()
        .unwrap()
        .exit_ok()
        .unwrap();

    let Output { status, stdout, stderr } = Command::new(&executable_filename)
        .args(cmdline_arguments.iter().map(Borrow::<str>::borrow))
        .output()
        .unwrap();
    let stdout = std::str::from_utf8(&stdout).unwrap();
    let stderr = std::str::from_utf8(&stderr).unwrap();
    let actual_exit_code = status.code().unwrap_or_else(
        #[cfg(unix)]
        || {
            panic!(
                "process `{}` died with {status}",
                executable_filename.display(),
            )
        },
        #[cfg(not(unix))]
        || unreachable!("`ExitStatus::code()` can only be `None` on Unix"),
    );

    assert_eq!(
        expected_return_code, actual_exit_code,
        "test program did not exit with expected return code {expected_return_code}",
    );

    assert_eq!(
        stdout, expected_output,
        "output on stdout (left) did not match expected output (right)",
    );
    assert_eq!(stderr, "", "no output on stderr is expected");
}
