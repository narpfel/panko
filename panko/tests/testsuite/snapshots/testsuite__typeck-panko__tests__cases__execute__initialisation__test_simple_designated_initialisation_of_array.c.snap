---
source: panko/tests/testsuite/main.rs
info:
  program: panko
  args:
    - "--print=typeck"
    - "--stop-after=typeck"
    - panko/tests/cases/execute/initialisation/test_simple_designated_initialisation_of_array.c
---
success: true
exit_code: 0
----- stdout -----
(translation-unit
   (declaration printf~0 `fn(∅: ptr<char const>, ...) -> int` ∅)
   (function-definition
      f~1 `fn(x: int) -> int`
      (params
         x~2 `int`)
      (compound-statement
         (definition s~3 `array<char const; (constexpr 4)>`
            (braced
               (subobject +0 (truncate (37 `int`) `char`))
               (subobject +1 (truncate (100 `int`) `char`))
               (subobject +2 (truncate (10 `int`) `char`))
               (subobject +3 (truncate (0 `int`) `char`))))
         (expression
            (call (addressof (printf~0 `fn(∅: ptr<char const>, ...) -> int`) `ptr<fn(∅: ptr<char const>, ...) -> int>`)
               (noop-type-conversion (addressof (s~3 `array<char const; ∅>`) `ptr<array<char const; ∅>>`) `ptr<char const>`)
               (x~2 `int`)
               `int`))
         (return (x~2 `int`))))
   (function-definition
      main~4 `fn(∅: int, argv: ptr<ptr<char>>) -> int`
      (params
         main.unnamed_parameter.0~5 `int`
         argv~6 `ptr<ptr<char>>`)
      (compound-statement
         (definition xs~7 `array<int; (constexpr 43)>`
            (braced
               (subobject +0
                  (call (addressof (f~1 `fn(x: int) -> int`) `ptr<fn(x: int) -> int>`)
                     (1 `int`)
                     `int`))
               (subobject +168
                  (call (addressof (f~1 `fn(x: int) -> int`) `ptr<fn(x: int) -> int>`)
                     (2 `int`)
                     `int`))
               (subobject +108
                  (call (addressof (f~1 `fn(x: int) -> int`) `ptr<fn(x: int) -> int>`)
                     (3 `int`)
                     `int`))
               (subobject +112
                  (call (addressof (f~1 `fn(x: int) -> int`) `ptr<fn(x: int) -> int>`)
                     (4 `int`)
                     `int`))
               (subobject +116
                  (call (addressof (f~1 `fn(x: int) -> int`) `ptr<fn(x: int) -> int>`)
                     (5 `int`)
                     `int`))
               (subobject +0
                  (call (addressof (f~1 `fn(x: int) -> int`) `ptr<fn(x: int) -> int>`)
                     (42 `int`)
                     `int`))))
         (definition s~8 `ptr<char const>` (noop-type-conversion (deref (ptr-add (argv~6 `ptr<ptr<char>>`) (sign-extend (1 `int`) `unsigned long`) `ptr<ptr<char>>`) `ptr<char>`) `ptr<char const>`))
         (expression
            (call (addressof (printf~0 `fn(∅: ptr<char const>, ...) -> int`) `ptr<fn(∅: ptr<char const>, ...) -> int>`)
               (s~8 `ptr<char const>`)
               (deref (ptr-add (noop-type-conversion (addressof (xs~7 `array<int; ∅>`) `ptr<array<int; ∅>>`) `ptr<int>`) (sign-extend (0 `int`) `unsigned long`) `ptr<int>`) `int`)
               `int`))
         (expression
            (call (addressof (printf~0 `fn(∅: ptr<char const>, ...) -> int`) `ptr<fn(∅: ptr<char const>, ...) -> int>`)
               (s~8 `ptr<char const>`)
               (deref (ptr-add (noop-type-conversion (addressof (xs~7 `array<int; ∅>`) `ptr<array<int; ∅>>`) `ptr<int>`) (sign-extend (1 `int`) `unsigned long`) `ptr<int>`) `int`)
               `int`))
         (expression
            (call (addressof (printf~0 `fn(∅: ptr<char const>, ...) -> int`) `ptr<fn(∅: ptr<char const>, ...) -> int>`)
               (s~8 `ptr<char const>`)
               (deref (ptr-add (noop-type-conversion (addressof (xs~7 `array<int; ∅>`) `ptr<array<int; ∅>>`) `ptr<int>`) (sign-extend (26 `int`) `unsigned long`) `ptr<int>`) `int`)
               `int`))
         (expression
            (call (addressof (printf~0 `fn(∅: ptr<char const>, ...) -> int`) `ptr<fn(∅: ptr<char const>, ...) -> int>`)
               (s~8 `ptr<char const>`)
               (deref (ptr-add (noop-type-conversion (addressof (xs~7 `array<int; ∅>`) `ptr<array<int; ∅>>`) `ptr<int>`) (sign-extend (27 `int`) `unsigned long`) `ptr<int>`) `int`)
               `int`))
         (expression
            (call (addressof (printf~0 `fn(∅: ptr<char const>, ...) -> int`) `ptr<fn(∅: ptr<char const>, ...) -> int>`)
               (s~8 `ptr<char const>`)
               (deref (ptr-add (noop-type-conversion (addressof (xs~7 `array<int; ∅>`) `ptr<array<int; ∅>>`) `ptr<int>`) (sign-extend (28 `int`) `unsigned long`) `ptr<int>`) `int`)
               `int`))
         (expression
            (call (addressof (printf~0 `fn(∅: ptr<char const>, ...) -> int`) `ptr<fn(∅: ptr<char const>, ...) -> int>`)
               (s~8 `ptr<char const>`)
               (deref (ptr-add (noop-type-conversion (addressof (xs~7 `array<int; ∅>`) `ptr<array<int; ∅>>`) `ptr<int>`) (sign-extend (29 `int`) `unsigned long`) `ptr<int>`) `int`)
               `int`))
         (expression
            (call (addressof (printf~0 `fn(∅: ptr<char const>, ...) -> int`) `ptr<fn(∅: ptr<char const>, ...) -> int>`)
               (s~8 `ptr<char const>`)
               (deref (ptr-add (noop-type-conversion (addressof (xs~7 `array<int; ∅>`) `ptr<array<int; ∅>>`) `ptr<int>`) (sign-extend (30 `int`) `unsigned long`) `ptr<int>`) `int`)
               `int`))
         (expression
            (call (addressof (printf~0 `fn(∅: ptr<char const>, ...) -> int`) `ptr<fn(∅: ptr<char const>, ...) -> int>`)
               (s~8 `ptr<char const>`)
               (deref (ptr-add (noop-type-conversion (addressof (xs~7 `array<int; ∅>`) `ptr<array<int; ∅>>`) `ptr<int>`) (sign-extend (41 `int`) `unsigned long`) `ptr<int>`) `int`)
               `int`))
         (expression
            (call (addressof (printf~0 `fn(∅: ptr<char const>, ...) -> int`) `ptr<fn(∅: ptr<char const>, ...) -> int>`)
               (s~8 `ptr<char const>`)
               (deref (ptr-add (noop-type-conversion (addressof (xs~7 `array<int; ∅>`) `ptr<array<int; ∅>>`) `ptr<int>`) (sign-extend (42 `int`) `unsigned long`) `ptr<int>`) `int`)
               `int`)))))

----- stderr -----
